src/kafka/kafka_clients.clj

```diff
@@ -0,0 +1,170 @@
+(ns kafka.kafka-clients
+  (:import (java.net InetAddress)
+           (org.apache.kafka.clients.producer KafkaProducer ProducerRecord)
+           (org.apache.kafka.common.serialization StringSerializer StringDeserializer)
+           (org.apache.kafka.clients.consumer KafkaConsumer)
+           (java.time Duration)))
+
+(defn build-producer-config-atom
+  []
+  (atom {"value.serializer"  StringSerializer
+         "key.serializer"    StringSerializer
+         "client.id"         (.getHostName (InetAddress/getLocalHost))
+         "bootstrap.servers" "localhost:9092"
+         "acks"              "all"}))
+
+(defn build-consumer-config-atom
+  []
+  (atom {"group.id"           "My-Group"
+         "key.deserializer"   StringDeserializer
+         "value.deserializer" StringDeserializer
+         "client.id"          (.getHostName (InetAddress/getLocalHost))
+         "bootstrap.servers"  "localhost:9092"}))
+
+(defn build-producer-atom
+  [config]
+  (let [config (deref config)]
+    (atom (new KafkaProducer config))))
+
+(defn build-consumer-atom
+  [config]
+  (let [config   (deref config)
+        consumer (new KafkaConsumer config)]
+    (.subscribe consumer ["demo"])
+    (atom consumer)))
```

Comment:
fns are unnecessary to build atoms. You should have fns to manipulate atoms, not generate them

Reply:
The atom should be `def` in the top level, just like `(def producer (atom (build-producer config)))`?